common/djangoapps/third_party_auth/pipeline.py

```diff
@@ -709,3 +713,72 @@ def associate_by_email_if_login_api(auth_entry, backend, details, user, current_
             # email address and the legitimate user would now login to the illegitimate
             # account.
             return association_response
+
+
+def user_details_force_sync(auth_entry, strategy, details, user=None, *args, **kwargs):
+    """
+    Update normally protected user details using data from provider.
+
+    This step in the pipeline is akin to `social_core.pipeline.user.user_details`, which updates
+    the user details but has an unconfigurable protection over updating the username & email, and
+    is unable to update information such as the user's full name which isn't on the user model, but
+    rather on the user profile model.
+
+    Additionally, because the email field is normally used to login, if the email is changed by this
+    forced synchronization, we send an email to both the old and new emails, letting the user know.
+
+    This step is controlled by the `sync_learner_profile_data` flag on the provider's configuration.
+    """
+    current_provider = provider.Registry.get_from_pipeline({'backend': strategy.request.backend.name, 'kwargs': kwargs})
+    if user and current_provider.sync_learner_profile_data:
+        # Keep track of which incoming values get applied.
+        changed = {}
+
+        # Map each incoming field from the provider to the name on the user model (by default, they always match).
+        field_mapping = {field: (user, field) for field in details.keys() if hasattr(user, field)}
+
+        # This is a special case where the field mapping should go to the user profile object and not the user object,
+        # in some cases with differing field names (i.e. 'fullname' vs. 'name').
+        field_mapping.update({
+            'fullname': (user.profile, 'name'),
+            'country': (user.profile, 'country'),
+        })
+
+        for provider_field, (model, field) in field_mapping.items():
+            provider_value = details.get(provider_field)
+            current_value = getattr(model, field)
+            if provider_value is not None and current_value != provider_value:
+                changed[provider_field] = current_value
+                setattr(model, field, provider_value)
+
+        if changed:
+            logger.info(
+                "User with ID [%s] performed SSO through [%s] who synchronizes profile data, and the "
+                "following fields were changed: %s", user.id, current_provider.name, changed.keys(),
+            )
+            strategy.storage.user.changed(user)
```

Comment:
@UmanShahzad What does this do? (Not suggesting it's obsolete, just curious -- it looks like something that the auth pipeline would require, but I don't know any details.)

Reply:
It literally just calls `user.save()`, but I'm using it because that's also what the upstream code calls for the `user_details` function, so maybe good to keep it that way.

Reply:
@UmanShahzad OK, that makes sense. Looking at this line, though, it's not really clear that its main purpose is to call `user.save()`. Let's add an inline comment that mentions this, something like:

```python
            # Save changes to user and user.profile models
            strategy.storage.user.changed(user)
            user.profile.save()
```

Reply:
Done! :+1: